### ISSUE TITLE
CoreFX System.Security.Cryptography.X509Certificates namespace

### ISSUE THREAD

**Issue by marek-safar** [mono/mono#9216]
## Types in System.dll to import

- [ ] CertificateRequest
- [ ] OpenFlags
- [ ] PublicKey
- [ ] StoreLocation
- [ ] StoreName
- [ ] X500DistinguishedName
- [ ] X500DistinguishedNameFlags
- [ ] X509BasicConstraintsExtension
- [ ] X509Certificate2
- [ ] X509Certificate2Collection
- [ ] X509Certificate2Enumerator
- [ ] X509CertificateCollection
- [ ] X509Chain
- [ ] X509ChainElement
- [ ] X509ChainElementCollection
- [ ] X509ChainElementEnumerator
- [ ] X509ChainPolicy
- [ ] X509ChainStatus
- [ ] X509ChainStatusFlags
- [ ] X509EnhancedKeyUsageExtension
- [ ] X509Extension
- [ ] X509ExtensionCollection
- [ ] X509ExtensionEnumerator
- [ ] X509FindType
- [ ] X509IncludeOption
- [ ] X509KeyUsageExtension
- [ ] X509KeyUsageFlags
- [ ] X509NameType
- [ ] X509RevocationFlag
- [ ] X509RevocationMode
- [ ] X509Store
- [ ] X509SubjectKeyIdentifierExtension
- [ ] X509SubjectKeyIdentifierHashAlgorithm
- [ ] X509VerificationFlags

## Required Dependencies
- [ ] AppleCrypto PAL #9417
- [ ] SafePasswordHandle #9416

**Comment by baulig** [comment]
**IMPORTANT**  Prior to working on any of these classes, please talk to me (Martin Baulig) first, so we can discuss a few things and make sure it is heading into the right direction.

Most importantly, we need to make a plan on how to proceed before writing any code.  This will require some research and discussion among the team as well.

- [ ] CoreFX and OpenSSL
CoreFX currently uses OpenSSL, which is unacceptable.
Can we either convince them to switch to boringssl or can we port their openssl PAL to boringssl?
- [ ] Which PAL do we want to use on iOS and what is our plan about missing features?  Several of the native crypto and certificate APIs are not supported on iOS, do we want to throw `PlatformNotSupportedException` or do we need some workaround?
- [ ] Does CoreFX have a native PAL for X.509 extensions for all platforms that we care about?  Do we need a managed implementation for some platforms?  If so, could we possibly contribute that to CoreFX?

**Comment by filipnavara** [comment]
I spent some time on a proof-of-concept to get all the CoreFX code integrated in Mono. The code is not production ready in any way, but it was good enough to figure out some issues that are going to be hit and to get the unit tests running on macOS. Having this experience I want to weigh in on the above questions and add few more.

### PAL

> CoreFX currently uses OpenSSL, which is unacceptable.
> Can we either convince them to switch to boringssl or can we port their openssl PAL to boringssl?

It seems rather straightforward to get the System.Native.Cryptography code compiling with BoringSSL. Testing and maintaining could be another issue, but the initial port should be trivial. Other open source projects use the `#if OPENSSL_IS_BORINGSSL` C preprocessor definition. I think the same approach could be used here.

> Which PAL do we want to use on iOS and what is our plan about missing features? Several of the native crypto and certificate APIs are not supported on iOS, do we want to throw PlatformNotSupportedException or do we need some workaround?

Do you have a list of APIs that are not supported? I think it would make sense to use the native Apple backend and just throw PlatformNotSupportedException when necessary. In fact, CoreFX already throws it way too many times on macOS so it would be consistent at least (https://github.com/dotnet/corefx/blob/master/Documentation/architecture/cross-platform-cryptography.md).

> Does CoreFX have a native PAL for X.509 extensions for all platforms that we care about? Do we need a managed implementation for some platforms? If so, could we possibly contribute that to CoreFX?

The code uses managed implementation on all non-Windows systems that should be good enough. Some behaviour differs from Mono/NetFX, but it's rather trivial to fix. 

> To summarize - do we need a managed implementation for anything or can we go fully PAL/native? And if we do, could these possibly be contributed to CoreFX, so we won't have to maintain them.

We may need a managed implementation if we care about exact output of some formatting functions (such as `X509Extension.Format`). The rest seems to just work (tm).

### Mono.Security

> Can we fully retire the old Mono.Security implementations of the certificate classes and move to CoreFX. I do not want to see any major code additions or changes to these classes before this is addressed and would ideally prefer any new managed implementations to be contributed to CoreFX.

Short answer: Maybe, eventually. It very much depends on how we decide to deal with `X509Certificate` in corlib. The issue is explained below.

I would personally very much like to retire it since it's quite buggy.

> Once Legacy TLS is gone, several other classes will be retired as well. This includes for instance X509Certificate2ImplMono and X509ChainImplMono.

Nice! That's good to know. `X509ChainImplMono` is quite incomplete and was one problematic piece of code to deal with.

> We should also investigate whether there are any pieces that we could possibly contribute to them.

Probably none. They're missing support for Authenticode and couple of other things, but new libraries in CoreFX make the Mono.Security code obsolete. For example there's PEReader class that reduces the Authenticode extraction to less than 10 lines. 

### Mono TLS Provider

For my proof of concept I basically hacked the code to always recreate the Apple/BTLS certificate from `X509Certificate.GetRawData()`. Obviously in production this has to be done differently. The Apple APIs seem to support everything that is necessary, although sometimes in quite a round-about way (which is already done in CoreFX code base).

### "The implementation questions"

- Do we want to support more than one X509/TLS backend for one platform? Mono currently builds both Apple TLS/certificate code on macOS. CoreFX code base currently doesn't support it, but it's definitely possible to fix (https://github.com/dotnet/corefx/issues/30606).

- Dependencies! Dependencies! Dependencies! While most of the classes reside in System.dll there's one big exception - `X509Certificate`. Since it's inevitably part of the implementation (`X509Certificate2` inherits from it) it can't be separated from the rest. The whole PAL layer depends on types found in System.dll though (and one type from System.Core, but that's not really a blocker). I found three possible ways handle this situation:
   1. Split `ICertificatePal` in CoreFX into `ICertificatePalV1` and `ICertificatePal : ICertificateV1`. `ICertificatePalV1` references only mscorlib types and contains only the members required by `X509Certificate`. Then X509Certificate.cs can be compiled into mscorlib, but still needs some PAL implementation. This PAL implementation could be implemented using the old Mono.Security code and be self-contained. _Pros:_ mscorlib remains self-contained _Cons:_ Mono.Security lives on, requires some minor changes to the CoreFX (though small enough to make it viable)
   2. Same as 1. except that the PAL is implemented as a small shim that loads System.dll and uses the PAL from there. _Pros:_ Mono.Security can go away, the PAL layer exists only once _Cons:_ mscorlib is not self-contained, would it work on AOT platforms?
   3. Type-forward X509Certificate from mscorlib to System.dll. Is that even possible?

- `X509Certificate[2]` in CoreFX is now "immutable" to some extent and all the `Import` methods are deprecated and throw an exception. Do we want to carry on this behaviour from CoreFX? Do we want to do what Mono does now? It's relatively easy change code-wise, but may not be easy to upstream.

- How do we deal with certificate types that are not natively supported on a system, but which were previously supported by Mono? There's an open issue on this: https://github.com/dotnet/corefx/issues/2910

**Comment by baulig** [comment]
We can possibly get around the mscorlib / System split via some dependency injection.  I actually thought about something like that in the past, but didn't have the time to actually look into it.

If we could somehow get the code in `mscorlib.dll` invoke a method from `System.dll`, then we could solve it via abstract classes / interfaces.

**Comment by filipnavara** [comment]
I do have a prototype of exactly that. It works like this:
- `ICertificatePal` from CoreFX is split into `ICertificatePalV1` and `ICertificatePal : ICertificatePalV1`.
- `X509Certificate` is modified (~ 5 lines) to use `ICertificatePalV1`.
- mscorlib declares `IStorePalShim` and `ICertificatePalShim` that mirror 2 static methods of `StorePal` and 4 static methods of `CertificatePal`.
- mscorlib contains two small `StorePal` and `CertificatePal` classes that expose the same static methods, but obtain them using reflection from System.dll (`Assembly.Load`, `Assembly.GetType`, `Activator.CreateInstance`).
- System contains additional `StorePalShim : IStorePalShim` and `CertificatePal : ICertificatePalShim` classes.

It does work on desktop reasonably well. I am not sure whether it would work on FullAOT platforms.

**Comment by filipnavara** [comment]
Tracking my BoringSSL compilation changes for CoreFX here: https://github.com/filipnavara/corefx/tree/boringssl-1

Built using `OPENSSL_ROOT_DIR=/usr/local/opt/boringssl ./build-native.sh`.

Quite a few things are missing (or possibly outright wrong), requires up-to-date BoringSSL as it depends on some APIs that were added only 2 weeks ago. Fails couple of tests on ECC cryptography, some of them because of missing "seed" support, some of them I didn't analyze yet. Only tested compilation on macOS so far, planning to move to Linux and fixing the X.509 code.

**Comment by baulig** [comment]
I have mostly figured out the dependency injection problem - see my new PR https://github.com/mono/mono/pull/9369.

The next step is to move the native `X509CertificateImplApple` from corlib into System and use `X509Certificate2` - I will take care of that shortly.

**Comment by baulig** [comment]
I am making quite some good progress on cleaning up the certificate code and am starting to get really happy with it.  In my latest working tree, I moved all the `X509CertificateImpl` implementations from corlib into System - this is hooked up via the `X509Helper` / `INativeCertificateHelper` dependency injection.  This mechanic will need some testing on mobile.

The implementation then actually uses `X509Certificate2Impl` (which uses the System APIs) for both BTLS and AppleTLS.

And then I started to use `CertificateData` from CoreFX (they just merged my PR which split that out into a separate file) for some of the code.

**Comment by filipnavara** [comment]
Looking forward to the clean up you are doing! It certainly seems like a sensible way to solve the dependency injection.

What's the plan with Mono's BoringSSL fork? I can backport the missing PKCS#7 and PKCS#12 functions. However I would rather see the fork to be updated with upstream fixes. I don't know what are the pitfalls of each of these approaches (endianness patches, etc.), so it would make sense to discuss it before writing any code.

**Comment by baulig** [comment]
We already planned to eventually upgrade it to the latest upstream, but just didn't get to actually doing it.

I can look into that shortly - maybe not next week because I will be out from Wed-Fri, but probably the week after.

**Comment by baulig** [comment]
The dependency injection problem should now be fixed, I got this worked out in my work tree.

I also did some major cleanups and refactoring to the certificate code, using CoreFX's implementation in quite a few places and I believe I found a _magical_ way of dealing with the "`CertificatePal` in corlib" problem.

The next step is to bring some of the apple crypto code from CoreFX - I created a task to keep track of this:
https://github.com/mono/mono/issues/9417.

My current code is too complex to land in one piece, so I'm going to refactor it into three smaller pieces:

1. The dependency injector.  This will also move `X509CertificateImplApple` from corlib into System, but without changing the implementation yet.
2. The AppleCrypto port and replacing some of the existing implementation with that.  I ran into the problem of having missing support for things like for instance PKCS12 import (without dealing with the annoying keychain mechanics).
3. My last refactoring patch.

I will be on vacation for the rest of the week, just wanted to give you a quick status update.

**Comment by filipnavara** [comment]
I cleaned-up a bit my CoreFX w/ BoringSSL branch and added OpenSSL 1.1 support to it (https://github.com/dotnet/corefx/pull/30574). There's few more issues that need to be dealt with, but I have a plan for those. Overall the patch is not very nice and it's likely not going to get in as-is. The current plan is to use it as a reference for a proper patch and use the good bits. If that would take too much time I can upstream as much as possible and make a small patch for the BoringSSL specific bits. The good news is that it is possible to use the native PAL w/ BoringSSL.

The PKCS12 stuff w/ AppleCrypto is a mess to deal with. In fact that's one big area where the Mono unit tests fail w/ the CoreFX implementation (https://github.com/dotnet/corefx/issues/24225, https://github.com/dotnet/corefx/issues/11046). I didn't have time to look into it yet, but I plan to at least make a list of issues I found using the Mono test suite.

Thanks for the update!

**Comment by baulig** [comment]
So this is the first batch of my ongoing work - it adds the dependency injector and moves all the `X509Certificate` implementation from corlib into System:  https://github.com/mono/mono/pull/9472.

The newly introduced `X509PalImpl` will eventually implement `IX509Pal` from CoreFX and the fallback implementation will go away.

I already moved `X509CertificateImplApple` from corlib into System - it will eventually derive from `X509Certificate2Impl` (with a new intermediate type that uses CoreFX's managed PAL instead of the fallback implementation) and use `SecCertificate` / `SecIdentity` instead of duplicating the P/Invokes.

Next step is to bring some of the native code and evaluate whether some of the internal `Mono.AppleTls` classes such as `SecCertificate` and `SecTrust` can be switched to CoreFX's PAL code.  I think I'll give this a try next to see whether that's cause any major trouble for mobile.

**Comment by baulig** [comment]
Here comes the next piece of work: https://github.com/mono/mono/pull/9524.

This refactors the AppleTls code to use `SafeSecCertificateHandle` and friends everywhere instead our `SecCertificate`.

We need to either add the `partial` modifier to CoreFX's version of those safe handles or move them into a file of their own (my preferred solution) because my new version requires an `IntPtr` constructor (and some hack to get `SecKeyRef` working properly).

**Comment by dhcgn** [comment]
Can we add the type `CertificateRequest` to this issue?

See in CoreFX:
[corefx/src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs](https://github.com/dotnet/corefx/blob/8d2407974c88b50c54f29ee20fd9deb8e6d00b1f/src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs)
